File: boj/GraphTheory/BOJ_3098.py
# 백준 3098 소셜네트워크
# 21.03.16
import sys
from functools import cmp_to_key
from collections import deque
from itertools import combinations
from itertools import permutations
rd = sys.stdin.readline
INF = int(1e9)

n, m = map(int, input().split())
graph = [[INF] * (n + 1) for _ in range(n + 1)]
cnt = m
days = [0] * (n + 1)
day = 1
for _ in range(m):
    a, b = map(int, rd().split())
    graph[a][b] = graph[b][a] = 0

while cnt < n*(n-1)//2:
    for k in range(1, n + 1):
        for i in range(1, n + 1):
            for j in range(1, n + 1):
                if i == j or i == k or j == k: continue
                if graph[i][j] == INF and graph[i][k] < day and graph[k][j] < day:
                    graph[i][j] = graph[j][i] = day
                    cnt += 1
                    days[day] += 1
    day += 1

print(day - 1)
for i in range(1, day):
    print(days[i])

File: boj/GraphTheory/BOJ_3098.swift
// 백준 3098 소셜네트워크
// 21.03.19

import Foundation

let input = readLine()!.split(separator: " ").map { Int(String($0))! }
let n = input[0]
let m = input[1]
let INF = Int(1e9)

var days = Array(repeating: 0, count: n + 1)
var graph = Array(repeating: Array(repeating: INF, count: n + 1), count: n + 1)

for _ in 0..<m {
  let line = readLine()!.split(separator: " ").map { Int(String($0))! }
  graph[line[0]][line[1]] = 0
  graph[line[1]][line[0]] = 0
}

var cnt = m
var day = 1

while cnt < n*(n-1)/2 {
  for o in stride(from: 1, through: n, by: 1) {
    for i in stride(from: 1, through: n, by: 1) {
      for j in stride(from: 1, through: n, by: 1) {
        if i == j || i == o || j == o { continue }
        if graph[i][j] == INF && graph[i][o] < day && graph[o][j] < day {
          graph[i][j] = day
          graph[j][i] = day
          days[day] += 1
          cnt += 1
        }
      }
    }
  }
  day += 1
}

print(day - 1)
for i in stride(from: 1, to: day, by: 1) { print(days[i]) }
